### PR TITLE
PATRON2020-11 Gateway

### PR DESCRIPTION
**Pull request by SzymonFed**

### CODE REVIEW

File: gateway/gateway.cpp [hunk 1]
```diff
@@ -0,0 +1,28 @@
+#include "dependencies/crow_all.h"
+#include "dependencies/json.hpp"
+
+int main()
+{
+    crow::SimpleApp gateway;
+    
+
+    CROW_ROUTE(gateway, "/hello")([](){
+        return "Hello world";
```
Comment (Siwy160) [inline gateway/gateway.cpp]: Hardcoded literal.

File: gateway/dashboard.json [hunk 2]
```diff
@@ -0,0 +1,68 @@
+{
+"temperatureSensors": [
+{
+"id": 0,
```
Comment (Siwy160) [inline gateway/dashboard.json]: Json file not properly formatted. Add tab intends.

Comment (SzymonFed) [reply gateway/dashboard.json]: done

File: gateway/dashboard.json [hunk 3]
```diff
@@ -0,0 +1,68 @@
+{
+    "temperatureSensors": [
+        {
+            "id": 0,
```
Comment (Siwy160) [inline gateway/dashboard.json]: Each id should be unique number starting from 0.

File: gateway/dashboard.json [hunk 4]
```diff
@@ -0,0 +1,68 @@
+{
+    "temperatureSensors": [
+        {
+            "id": 0,
+            "type": "string",
```
Comment (Siwy160) [inline gateway/dashboard.json]: It should be descriptive type of each sensor. Like:
temperature
window
windowBlind
smokeSensor
RFIdSensor HVACStatus
light etc.

File: gateway/dashboard.json [hunk 5]
```diff
@@ -0,0 +1,68 @@
+{
+    "temperatureSensors": [
+        {
+            "id": 0,
+            "type": "string",
+            "value": 0
```
Comment (Siwy160) [inline gateway/dashboard.json]: Value should be a number from correct out range from sensor.
0 for temperature is actually fine. But lets go with some more natural value. Eg. 24

File: gateway/dashboard.json [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+        }
+    ],
+    "RFIDSensors": [
+        {
+            "id": 0,
+            "type": "string",
+            "lastTag": {
+                "id": 0,
+                "type": "string",
+                "timestamp": 0
```
Comment (Siwy160) [inline gateway/dashboard.json]: Put some actual timestamp here please.
eg. 1584616475765 (source: https://currentmillis.com/ )

File: gateway/dashboard.json [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+    "HVACStatus": {
+        "type": "string",
+        "heating": false,
+        "cooling": false
+    },
+    "HVACRooms": [
+        {
+            "id": 0,
+            "type": "string",
+            "heating-temperature": 0,
```
Comment (Siwy160) [inline gateway/dashboard.json]: More real value please. eg 22

File: gateway/dashboard.json [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+            "cooling-temperature": 0,
```
Comment (Siwy160) [inline gateway/dashboard.json]: More real value please. eg 26

File: gateway/dashboard.json [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+            "hysteresis": 0,
```
Comment (Siwy160) [inline gateway/dashboard.json]: More real value. eg 2

File: gateway/dashboard.json [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+            "temperatureSensorId": 0,
```
Comment (Siwy160) [inline gateway/dashboard.json]: Put actual id of one of the sensors returned.

File: gateway/dashboard.json [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+            "windowSensorIds": [
+                0
```
Comment (Siwy160) [inline gateway/dashboard.json]: Put actual id of one of the sensors.

File: gateway/dashboard.json [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+            "type": "string",
+            "isSmokeDetected": false
+        }
+    ],
+    "HVACStatus": {
+        "type": "string",
+        "heating": false,
+        "cooling": false
+    },
+    "HVACRooms": [
```
Comment (Siwy160) [inline gateway/dashboard.json]: would be nice to have 2 rooms.

File: gateway/dashboard.json [hunk 13]
```diff
@@ -0,0 +1,68 @@
+{
+    "temperatureSensors": [
+        {
+            "id": 0,
+            "type": "string",
+            "value": 0
+        }
+    ],
+    "windowSensors": [
```
Comment (Siwy160) [inline gateway/dashboard.json]: Would be nice to have 3 sensors.
2 - one room
1 - second room

File: gateway/dashboard.json [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+        }
+    ],
+    "windowSensors": [
+        {
+            "id": 0,
+            "type": "string",
+            "status": "open"
+        }
+    ],
+    "windowBlinds": [
```
Comment (Siwy160) [inline gateway/dashboard.json]: We are going to have 2 blinds in total.
One for each room.

File: gateway/dashboard.json [hunk 15]
```diff
@@ -0,0 +1,68 @@
+{
+    "temperatureSensors": [
```
Comment (Siwy160) [inline gateway/dashboard.json]: At least 2 sensors please.
One for each hvac room.

File: gateway/gateway.cpp [hunk 16]
```diff
@@ -0,0 +1,28 @@
+#include "dependencies/crow_all.h"
+#include "dependencies/json.hpp"
+
+int main()
+{
+    crow::SimpleApp gateway;
+    
+
```
Comment (Siwy160) [inline gateway/gateway.cpp]: One blank line is enough.

File: gateway/gateway.cpp (hunk 1 above)
Comment (Siwy160) [inline gateway/gateway.cpp]: Incorrect intends. Should be 8 spaces for this line. I've count 7 here ;)

File: gateway/gateway.cpp [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+    CROW_ROUTE(gateway, "/hello")([](){
+        return "Hello world";
+    });
+
+    CROW_ROUTE(gateway, "/dashboard").methods(crow::HTTPMethod::Get)([](){
+        
+        crow::json::wvalue x;
+        nlohmann::json j;
```
Comment (Siwy160) [inline gateway/gateway.cpp]: More descriptive names please?

File: gateway/gateway.cpp [hunk 18]
```diff
@@ -0,0 +1,28 @@
...
+        std::ifstream i ("dashboard.json");
+        i>>j;
+        i.close();
+        
+        x = crow::json::load(j.dump());
+
+        return crow::response(x);
+    });
+
+    gateway.port(8080).multithreaded().run();
```
Comment (Siwy160) [inline gateway/gateway.cpp]: We are going to use 42893. Please change it now, so we will avoid problems when setting it up.

Port could be exported somewhere. Or passed via some argument. Not sure how to solve it correctly.
You can leave a TODO and create a ticket on Jira for improvement.

File: gateway/dashboard.json (hunk 15 above)
Comment (SzymonFed) [reply gateway/dashboard.json]: done

File: gateway/gateway.cpp (hunk 18 above)
Comment (SzymonFed) [reply gateway/gateway.cpp]: Done. Defined port number in macro for now.

File: gateway/dashboard.json (hunk 3 above)
Comment (SzymonFed) [reply gateway/dashboard.json]: done

File: gateway/dashboard.json (hunk 4 above)
Comment (SzymonFed) [reply gateway/dashboard.json]: done

File: gateway/dashboard.json (hunk 5 above)
Comment (SzymonFed) [reply gateway/dashboard.json]: done

File: gateway/dashboard.json (hunk 7 above)
Comment (SzymonFed) [reply gateway/dashboard.json]: done

File: gateway/gateway.cpp (hunk 16 above)
Comment (SzymonFed) [reply gateway/gateway.cpp]: done

File: gateway/gateway.cpp (hunk 1 above)
Comment (SzymonFed) [reply gateway/gateway.cpp]: done

File: gateway/gateway.cpp [hunk 19]
```diff
@@ -0,0 +1,29 @@
+#include "dependencies/crow_all.h"
+#include "dependencies/json.hpp"
```
Comment (cichy1024) [inline gateway/gateway.cpp]: Is there any good reason you don't use JSON parser from Crow?

Comment (SzymonFed) [reply gateway/gateway.cpp]: Couldn't find any good documentation about Crow's JSON parser, if there's any. Nlohmann was easier and faster to understand :)

**Pull request merged**